src/shared_utilities/utils.jl

```diff
@@ -494,46 +500,84 @@ function isdivisible(
 end
 
 """
-    count_nans_state(state, mask::ClimaCore.Fields.Field = nothing, verbose = false)
+    call_count_nans_state(state, mask::ClimaCore.Fields.Field = nothing, verbose = false)
 
-Count the number of NaNs in the state, e.g. the FieldVector given by
+This calls the function which counts the number of NaNs in the state, e.g. the FieldVector given by
 `sol.u[end]` after calling `solve`. This function is useful for
-debugging simulations to determine quantitatively if a simulation is stable.
+debugging simulations to determine quantitatively if a simulation is stable; this is only
+for use in non-column simulations.
 
 If this function is called on a FieldVector, it will recursively call itself
 on each Field in the FieldVector. If it is called on a Field, it will count
-the number of NaNs in the Field and produce a warning if any are found.
+the number of NaNs in the Field and produce a warning if any are found; this behavior
+is different for 2d and 3d fields, which is why we dispatch on the `space`.
 
 If a ClimaCore Field is provided as `mask`, the function will only count NaNs
 in the state variables where the mask is 1. This is intended to be used with
 the land/sea mask, to avoid counting NaNs over the ocean. Note this assumes
-the mask is 1 over land and 0 over ocean.
+the mask is 1 over land and 0 over ocean, and that the mask is a 2D field.
 
 The `verbose` argument toggles whether the function produces output when no
 NaNs are found.
 """
-function count_nans_state(
+function call_count_nans_state(
     state::ClimaCore.Fields.FieldVector;
     mask = nothing,
     verbose = false,
 )
     for pn in propertynames(state)
         state_new = getproperty(state, pn)
         @info "Checking NaNs in $pn"
-        count_nans_state(state_new; mask, verbose)
+        call_count_nans_state(state_new; mask, verbose)
     end
     return nothing
 end
 
+function call_count_nans_state(
+    state::ClimaCore.Fields.Field;
+    mask = nothing,
+    verbose = false,
+)
+    if typeof(axes(state)) <: ClimaCore.Spaces.AbstractSpectralElementSpace
+        return count_nans_state(state; mask = mask, verbose = verbose)
+    elseif typeof(axes(state)) <: ClimaCore.Spaces.ExtrudedFiniteDifferenceSpace
+        return count_nans_state(
+            ClimaLand.top_center_to_surface(state);
+            mask = mask,
+            verbose = verbose,
+        )
```

Comment:
Instead of the if/else here, we could pass `axes(state)` to `count_nans_state`, since now we know `state` is a Field. Then we can dispatch on the space type and call `top_center_to_surface(state)` in the 3D case

Reply:
ok! could you clarify how this is different from what I had before?

Reply:
If we call count_nans_state with the space, we end up having to duplicate the counting code. But we can call call_count_nans_state with the space, and then call count_nans_state. I guess this is a combo of the previous + current. let me try it!